### ISSUE TITLE
LDAP server fails in JRuby with DER decoding exception.

### ISSUE THREAD

**Issue by mjstrasser** [inscitiv/ruby-ldapserver#3]
I have an LDAP server stub we use for performance-testing an application. The server works in MRI Ruby on Windows and Linux but fails in JRuby. We want to run the stub on Solaris, for which we have only JRuby available.

I have tried with JRuby versions 1.7.9 and 1.7.18 but connections fail with the same exception (this is from JRuby 1.7.18 on Windows):

    E, [2015-02-24T09:39:51.009000 #4436] ERROR -- 0:0:0:0:0:0:0:1: unable to decode object: org.bouncycastle.asn1.DERApplicationSpecific@492ea2e5 (org.bouncycastle.asn1.DERApplicationSpecific): org/jruby/ext/openssl/ASN1.java:1137:in `decode'

My test is an LDAP search and the first network request (BIND) fails. I traced it using the debugger in IntelliJ IDEA + Ruby plugin.

In `ruby-ldapserver` 0.5.1 the exception is thrown in the call to `OpenSSL::ASN1#decode` in `LDAP::Server::Connection#handle_requests`, line 102:

- In MRI Ruby `blk` is reported as a 26-byte array with ASCII-8BIT encoding. It decodes successfully and the debugger shows its contents as:
```
""0\x18\x02\x01\x01`\x13\x02\x01\x03\x04\to=Telstra\x80\x03ess""
```
- In JRuby `blk` is also reported as a 26-byte array with ASCII-8BIT encoding containing the same values _but the debugger shows its contents unescaped_. It does not decode successfully.

This side-by-side screenshot shows how the debugger displays the two byte arrays.

![ruby-ldapserver-blk-array](https://cloud.githubusercontent.com/assets/168173/6339790/3cc2cdc4-bc0d-11e4-87c3-b07b2e45ef5e.png)

`TCPSocket#read` in `LDAP::Server::Connection#ber_read` is returning bytes somehow differently in JRuby to MRI Ruby. Do you have any idea what this difference is and if there is a workaround I can use?

**Comment by dividedmind** [comment]
That's interesting. Frankly I haven't used JRuby much, but when porting to Ruby 1.9 I also encountered some encoding weirdness, see 6df0fa1090241f9bee46c6f81dac608108405ab1 -- I had to adjust tests for it. I suppose JRuby has its own version of the weirdness. Maybe throwing `force_encoding('ascii-8bit')` somewhere would fix it? Maybe the encoding it reports is spurious somehow.

Do the tests/spec catch this? If not, could you make a test case?

**Comment by dividedmind** [comment]
Ah, the problem seems to be that JRuby-OpenSSL doesn't handle Application Specific objects. I'm not sure how to solve this.

**Comment by dividedmind** [comment]
Specifically: there should be a case for `DERApplicationSpecific` after https://github.com/jruby/jruby-openssl/blob/master/src/main/java/org/jruby/ext/openssl/ASN1.java#L1085 -- looking essentially the same as for `ASNTaggedObject`, but setting `tag_class` to `APPLICATION` instead of `CONTEXT_SPECIFIC`. I don't have time to code and test it and report the bug upstream, but do let me know if you do.

**Comment by mjstrasser** [comment]
Thank you very much for digging further into it. I was unsure of how to debug across the Java–Ruby boundary in JRuby.

I will look into OpenSSL in JRuby when I have some time.

**Comment by mjstrasser** [comment]
Hi Rafał, I have implemented `ASN1.decodeObject` for `DERApplicationSpecific` as you suggested. I hope I got this right:
```
        if ( obj instanceof DERApplicationSpecific ) {
            final DERApplicationSpecific derAppSpecific = (DERApplicationSpecific) obj;
            IRubyObject val = decodeObject(context, ASN1, derAppSpecific.getObject());
            IRubyObject tag = runtime.newFixnum( derAppSpecific.getApplicationTag() );
            IRubyObject tag_class = runtime.newSymbol("APPLICATION");
            final RubyArray valArr = runtime.newArray(val);
            return ASN1.getClass("ASN1Data").callMethod(context, "new",
                new IRubyObject[] { valArr, tag, tag_class }
            );
        }
```
I now get this error:
```
undefined method `value' for nil:NilClass: /Users/Mike/dev/ruby/ruby-ldapserver/lib/ldap/server/operation.rb:195:in `do_bind'
```
The `protocolOp` object passed to `do_bind`’s `value` method returns an array with only one element, so the call to `value[1].value` throws the exception.

I deduce that more information needs to be read from the input stream into `protocolOp`. Are you able to guide me with this?

**Comment by dividedmind** [comment]
You're right; I'm looking at jruby-openssl and more and more think it's a bit too messy ;)

Anyway, I *think* what'd work is to interpret the data as a sequence, eg.
```java
final ASN1Sequence sequence = (ASN1Sequence) applicationSpecific.getObject(SEQUENCE);
final RubyArray valArr = decodeObjects(context, ASN1, sequence.getObjects());
```
instead of
```java
IRubyObject val = decodeObject(context, ASN1, derAppSpecific.getObject());
final RubyArray valArr = runtime.newArray(val);
```

I don't know if in general this is how ASN1 decoding is supposed to work, but in this case it seems to more or less do what's it's supposed to. Let me know how it fares.

**Comment by mjstrasser** [comment]
The bind seemd to work but the next step (a search) gets stuck. I have a no-op implementation of `simple_bind` and it works whether I provide bind credentials or not. Here is an example without:
```bash
ldapsearch -x -H ldap://127.0.0.1:1389/ -D "o=Telstra" -w pass -b "msisdn=61414284256,o=Telstra" "(objectclass=*)"
```
The first pass through the loop in `Connection#handle_requests` seems to be OK now, with `@binddn` and `@version` set correctly.

The second pass gets stuck in `Connection#ber_read` where line 58 never returns:
```ruby
    blk = io.read(2)
```
Any ideas?

**Comment by jbandhauer** [comment]
I'll admit to not having researched the specific of this issue much. But, I use ruby-ldap as an ldap client in jruby and they *seem* to have avoided this class of problems by implementing their own BER support:
https://github.com/ruby-ldap/ruby-net-ldap/tree/master/lib/net/ber  I wonder if this is sort of solution is viable from the server perspective. I can't commit to doing the work to try to make this work right now. But, I really would love to be able to integrate ruby-ldapserver into my jruby tests without having to manage it as a separate process.

**Comment by mjstrasser** [comment]
Yes, ruby-ldap do their own BER work. I needed to create an LDAP server stub using JRuby and discovered that ruby-ldap has enough server hooks for me to do that.

With that and eventmachine I was able to create a stub that performed well under load.
